From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mythiccal <cwood2014.cw@gmail.com>
Date: Tue, 22 Sep 2026 16:21:34 +0100
Subject: [PATCH] Parallel world ticking

Per-world tick thread checks and Bukkit tick-time accessors for SparklyPaper parallel world ticking.

diff --git a/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java b/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
index adb643a87ad754052c6097aeefc701e7b73f9cbb..36f76f97e3c1494be98c11b1bc6b6b063e3de171 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/IteratorSafeOrderedReferenceSet.java
@@ -127,7 +127,7 @@ public final class IteratorSafeOrderedReferenceSet<E> {
         }
     }
 
-    public boolean remove(final E element) {
+    public synchronized boolean remove(final E element) { // Pufferfish - Parallel world ticking
         final int index = this.indexMap.removeInt(element);
         if (index >= 0) {
             if (this.firstInvalidIndex < 0 || index < this.firstInvalidIndex) {
@@ -146,11 +146,11 @@ public final class IteratorSafeOrderedReferenceSet<E> {
         return false;
     }
 
-    public boolean contains(final E element) {
+    public synchronized boolean contains(final E element) { // Pufferfish - Parallel world ticking
         return this.indexMap.containsKey(element);
     }
 
-    public boolean add(final E element) {
+    public synchronized boolean add(final E element) { // Pufferfish - Parallel world ticking
         final int listSize = this.listSize;
 
         final int previous = this.indexMap.putIfAbsent(element, listSize);
diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
index 69cdd304d255d52c9b7dc9b6a33ffdb630b79abe..2405062586cd7afdb6fe670b32aa88c34998d01e 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -14,6 +14,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 public class TickThread extends Thread {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(TickThread.class);
+    public static final boolean HARD_THROW = !gg.pufferfish.pufferfish.PufferfishConfig.disableHardThrow; // Pufferfish - Parallel world ticking
 
     private static String getThreadContext() {
         return "thread=" + Thread.currentThread().getName();
@@ -26,14 +27,14 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final String reason) {
         if (!isTickThread()) {
             LOGGER.error("Thread failed main thread check: " + reason + ", context=" + getThreadContext(), new Throwable());
-            throw new IllegalStateException(reason);
+            if (HARD_THROW) throw new IllegalStateException(reason); // Pufferfish - Parallel world ticking
         }
     }
 
     public static void ensureTickThread(final Level world, final BlockPos pos, final String reason) {
         if (!isTickThreadFor(world, pos)) {
             final String ex = "Thread failed main thread check: " +
-                               reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos;
+                               reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + " - " + getTickThreadInformation(world.getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
@@ -42,7 +43,7 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final Level world, final BlockPos pos, final int blockRadius, final String reason) {
         if (!isTickThreadFor(world, pos, blockRadius)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + ", block_radius=" + blockRadius;
+                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + pos + ", block_radius=" + blockRadius + " - " + getTickThreadInformation(world.getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
@@ -60,7 +61,7 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final Level world, final int chunkX, final int chunkZ, final String reason) {
         if (!isTickThreadFor(world, chunkX, chunkZ)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + new ChunkPos(chunkX, chunkZ);
+                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", chunk_pos=" + new ChunkPos(chunkX, chunkZ) + " - " + getTickThreadInformation(world.getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
@@ -69,7 +70,7 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final Entity entity, final String reason) {
         if (!isTickThreadFor(entity)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", entity=" + EntityUtil.dumpEntity(entity);
+                reason + ", context=" + getThreadContext() + ", entity=" + EntityUtil.dumpEntity(entity) + " - " + getTickThreadInformation(entity.level().getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
@@ -78,7 +79,7 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final Level world, final AABB aabb, final String reason) {
         if (!isTickThreadFor(world, aabb)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", aabb=" + aabb;
+                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", aabb=" + aabb + " - " + getTickThreadInformation(world.getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
@@ -87,12 +88,67 @@ public class TickThread extends Thread {
     public static void ensureTickThread(final Level world, final double blockX, final double blockZ, final String reason) {
         if (!isTickThreadFor(world, blockX, blockZ)) {
             final String ex = "Thread failed main thread check: " +
-                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + new Vec3(blockX, 0.0, blockZ);
+                reason + ", context=" + getThreadContext() + ", world=" + WorldUtil.getWorldName(world) + ", block_pos=" + new Vec3(blockX, 0.0, blockZ) + " - " + getTickThreadInformation(world.getServer()); // Pufferfish - Parallel world ticking
             LOGGER.error(ex, new Throwable());
             throw new IllegalStateException(ex);
         }
     }
 
+    // Pufferfish start - Parallel world ticking
+    public static void ensureTickThread(final net.minecraft.server.level.ServerLevel world, final String reason) {
+        if (!isTickThreadFor(world)) {
+            LOGGER.error("Thread " + Thread.currentThread().getName() + " failed main thread check: " + reason + " @ world " + world.getWorld().getName() + " - " + getTickThreadInformation(world.getServer()), new Throwable());
+            if (HARD_THROW) throw new IllegalStateException(reason);
+        }
+    }
+
+    public static void ensureOnlyTickThread(final String reason) {
+        boolean isTickThread = isTickThread();
+        boolean isServerLevelTickThread = isServerLevelTickThread();
+        if (!isTickThread || isServerLevelTickThread) {
+            LOGGER.error("Thread " + Thread.currentThread().getName() + " failed main thread ONLY tick thread check: " + reason, new Throwable());
+            if (HARD_THROW) throw new IllegalStateException(reason);
+        }
+    }
+
+    public static void ensureTickThreadOrAsyncThread(final net.minecraft.server.level.ServerLevel world, final String reason) {
+        boolean isValidTickThread = isTickThreadFor(world);
+        boolean isAsyncThread = !isTickThread();
+        boolean isValid = isAsyncThread || isValidTickThread;
+        if (!isValid) {
+            LOGGER.error("Thread " + Thread.currentThread().getName() + " failed main thread or async thread check: " + reason + " @ world " + world.getWorld().getName() + " - " + getTickThreadInformation(world.getServer()), new Throwable());
+            if (HARD_THROW)
+                throw new IllegalStateException(reason);
+        }
+    }
+
+    public static String getTickThreadInformation(net.minecraft.server.MinecraftServer minecraftServer) {
+        StringBuilder sb = new StringBuilder();
+        Thread currentThread = Thread.currentThread();
+        sb.append("Is tick thread? ");
+        sb.append(currentThread instanceof TickThread);
+        sb.append("; Is server level tick thread? ");
+        sb.append(currentThread instanceof ServerLevelTickThread);
+        if (currentThread instanceof ServerLevelTickThread serverLevelTickThread) {
+            sb.append("; Currently ticking level: ");
+            if (serverLevelTickThread.currentlyTickingServerLevel != null) {
+                sb.append(serverLevelTickThread.currentlyTickingServerLevel.getWorld().getName());
+            } else {
+                sb.append("null");
+            }
+        }
+        sb.append("; Is iterating over levels? ");
+        sb.append(minecraftServer.isIteratingOverLevels);
+        sb.append("; Are we going to hard throw? ");
+        sb.append(HARD_THROW);
+        return sb.toString();
+    }
+
+    public static boolean isServerLevelTickThread() {
+        return Thread.currentThread() instanceof ServerLevelTickThread;
+    }
+    // Pufferfish end - Parallel world ticking
+
     public final int id; /* We don't override getId as the spec requires that it be unique (with respect to all other threads) */
 
     private static final AtomicInteger ID_GENERATOR = new AtomicInteger();
@@ -126,47 +182,71 @@ public class TickThread extends Thread {
         return false;
     }
 
+    // Pufferfish start - Parallel world ticking
     public static boolean isTickThreadFor(final Level world, final BlockPos pos) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final BlockPos pos, final int blockRadius) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final ChunkPos pos) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final Vec3 pos) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final int chunkX, final int chunkZ) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final AABB aabb) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final double blockX, final double blockZ) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final Vec3 position, final Vec3 deltaMovement, final int buffer) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final int fromChunkX, final int fromChunkZ, final int toChunkX, final int toChunkZ) {
-        return isTickThread();
+        return isTickThreadFor(world);
     }
 
     public static boolean isTickThreadFor(final Level world, final int chunkX, final int chunkZ, final int radius) {
-        return isTickThread();
+        return isTickThreadFor(world);
+    }
+
+    public static boolean isTickThreadFor(final Level world) {
+        if (Thread.currentThread() instanceof ServerLevelTickThread serverLevelTickThread) {
+            return serverLevelTickThread.currentlyTickingServerLevel == world;
+        } else return isTickThread();
     }
 
     public static boolean isTickThreadFor(final Entity entity) {
-        return isTickThread();
+        if (entity == null) {
+            return true;
+        }
+
+        return isTickThreadFor(entity.level());
+    }
+
+    public static class ServerLevelTickThread extends TickThread {
+        public ServerLevelTickThread(String name) {
+            super(name);
+        }
+
+        public ServerLevelTickThread(Runnable run, String name) {
+            super(run, name);
+        }
+
+        public net.minecraft.server.level.ServerLevel currentlyTickingServerLevel;
     }
+    // Pufferfish end - Parallel world ticking
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 716c3b17dc07eb2417dfb0084fe75dbd270f031a..c9aacefcc3492709431e9a9e8cb4ad55b95ee9e7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2961,4 +2961,11 @@ public final class CraftServer implements Server {
     public void allowPausing(final Plugin plugin, final boolean value) {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
+
+    // Pufferfish start - Parallel world ticking
+    @Override
+    public boolean isParallelWorldTickingEnabled() {
+        return gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking;
+    }
+    // Pufferfish end - Parallel world ticking
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f3f068011abd8c6585463169d50634bf42a09497..da05e363fd63d92e504ef47bef4f9007061bf5c8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -449,7 +449,13 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public boolean unloadChunkRequest(int x, int z) {
-        org.spigotmc.AsyncCatcher.catchOp("chunk unload"); // Spigot
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, x, z, "Cannot unload chunk asynchronously");
+        } else {
+            org.spigotmc.AsyncCatcher.catchOp("chunk unload"); // Spigot
+        }
+        // Pufferfish end - Parallel world ticking
         if (this.isChunkLoaded(x, z)) {
             this.world.getChunkSource().removeTicketWithRadius(TicketType.PLUGIN, new ChunkPos(x, z), 1);
         }
@@ -475,6 +481,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public boolean refreshChunk(int x, int z) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, x, z, "Cannot refresh chunk asynchronously"); // Pufferfish - Parallel world ticking
         ChunkHolder playerChunk = this.world.getChunkSource().chunkMap.getVisibleChunkIfPresent(ChunkPos.pack(x, z));
         if (playerChunk == null) return false;
 
@@ -525,7 +532,13 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public boolean loadChunk(int x, int z, boolean generate) {
-        org.spigotmc.AsyncCatcher.catchOp("chunk load"); // Spigot
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.getHandle(), x, z, "May not sync load chunks asynchronously");
+        } else {
+            org.spigotmc.AsyncCatcher.catchOp("chunk load"); // Spigot
+        }
+        // Pufferfish end - Parallel world ticking
         warnUnsafeChunk("loading a faraway chunk", x, z); // Paper
         ChunkAccess chunk = this.world.getChunkSource().getChunk(x, z, generate || isChunkGenerated(x, z) ? ChunkStatus.FULL : ChunkStatus.EMPTY, true); // Paper
 
@@ -738,6 +751,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public boolean generateTree(Location loc, TreeType type, BlockChangeDelegate delegate) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, loc.getX(), loc.getZ(), "Cannot generate tree asynchronously"); // Pufferfish - Parallel world ticking
         this.world.captureTreeGeneration = true;
         this.world.captureBlockStates = true;
         List<org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates;
@@ -840,6 +854,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     private boolean createExplosion(double x, double y, double z, float power, boolean setFire, boolean breakBlocks, Entity source, Consumer<net.minecraft.world.level.ServerExplosion> configurator) {
         // Paper end - expand explosion API
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, x, z, "Cannot create explosion asynchronously"); // Pufferfish - Parallel world ticking
         net.minecraft.world.level.Level.ExplosionInteraction explosionType;
         if (!breakBlocks) {
             explosionType = net.minecraft.world.level.Level.ExplosionInteraction.NONE; // Don't break blocks
@@ -898,6 +913,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public int getHighestBlockYAt(int x, int z, org.bukkit.HeightMap heightMap) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, x >> 4, z >> 4, "Cannot retrieve chunk asynchronously"); // Pufferfish - Parallel world ticking
         warnUnsafeChunk("getting a faraway chunk", x >> 4, z >> 4); // Paper
         // Transient load for this tick
         return this.world.getChunk(x >> 4, z >> 4).getHeight(CraftHeightMap.toNMS(heightMap), x, z);
@@ -913,6 +929,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     @Override
     public void setBiome(int x, int y, int z, Holder<net.minecraft.world.level.biome.Biome> biome) {
         BlockPos pos = new BlockPos(x, 0, z);
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, pos, "Cannot retrieve chunk asynchronously"); // Pufferfish - Parallel world ticking
         if (this.world.hasChunkAt(pos)) {
             net.minecraft.world.level.chunk.LevelChunk chunk = this.world.getChunkAt(pos);
 
@@ -1863,6 +1880,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public void sendGameEvent(Entity sourceEntity, org.bukkit.GameEvent gameEvent, Vector position) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.world, position.getX(), position.getZ(), "Cannot send game event asynchronously"); // Pufferfish - Parallel world ticking
         getHandle().gameEvent(sourceEntity != null ? ((CraftEntity) sourceEntity).getHandle(): null, net.minecraft.core.registries.BuiltInRegistries.GAME_EVENT.get(org.bukkit.craftbukkit.util.CraftNamespacedKey.toMinecraft(gameEvent.getKey())).orElseThrow(), org.bukkit.craftbukkit.util.CraftVector.toBlockPos(position));
     }
     // Paper end
@@ -2049,4 +2067,15 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Pufferfish start - Parallel world ticking
+    @Override
+    public long @NotNull [] getTickTimes() {
+        return this.world.tickTimes5s.getTimes();
+    }
+    @Override
+    public double getAverageTickTime() {
+        return this.world.tickTimes5s.getAverage();
+    }
+    // Pufferfish end - Parallel world ticking
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 72aacb573365c521609ce35af6b430f525edf3e0..a41f1b3de36a68e72d02d0a8caf0aab3d32e88c7 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -80,6 +80,11 @@ public class CraftBlock implements Block {
     }
 
     public net.minecraft.world.level.block.state.BlockState getBlockState() {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         return this.level.getBlockState(this.position);
     }
 
@@ -151,6 +156,11 @@ public class CraftBlock implements Block {
     }
 
     private void setData(final byte data, @net.minecraft.world.level.block.Block.UpdateFlags int flags) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, this.position, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         this.level.setBlock(this.position, CraftMagicNumbers.getBlock(this.getType(), data), flags);
     }
 
@@ -191,6 +201,11 @@ public class CraftBlock implements Block {
     }
 
     public static boolean setBlockState(LevelAccessor level, BlockPos pos, net.minecraft.world.level.block.state.BlockState newState, boolean applyPhysics) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, pos, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         int updateFlags = net.minecraft.world.level.block.Block.UPDATE_CLIENTS;
         if (applyPhysics) {
             updateFlags |= net.minecraft.world.level.block.Block.UPDATE_NEIGHBORS;
@@ -302,16 +317,31 @@ public class CraftBlock implements Block {
 
     @Override
     public Biome getBiome() {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         return this.getWorld().getBiome(this.getX(), this.getY(), this.getZ());
     }
 
     @Override
     public Biome getComputedBiome() {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         return this.getWorld().getComputedBiome(this.getX(), this.getY(), this.getZ());
     }
 
     @Override
     public void setBiome(Biome bio) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, this.position, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         this.getWorld().setBiome(this.getX(), this.getY(), this.getZ(), bio);
     }
 
@@ -361,6 +391,11 @@ public class CraftBlock implements Block {
 
     @Override
     public boolean isBlockFaceIndirectlyPowered(BlockFace face) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         Direction direction = blockFaceToNotch(face);
         Preconditions.checkArgument(direction != null, face + " is not a valid cartesian face");
         if (this.level.getMinecraftWorld().hasSignal(this.position, direction)) {
@@ -378,6 +413,11 @@ public class CraftBlock implements Block {
     @Override
     public int getBlockPower(BlockFace face) {
         Preconditions.checkArgument(face != null, "face cannot be null");
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
 
         net.minecraft.world.level.Level level = this.level.getMinecraftWorld();
         boolean searchSelfIncluded = face == BlockFace.SELF;
@@ -453,6 +493,11 @@ public class CraftBlock implements Block {
 
     @Override
     public boolean breakNaturally() {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, this.position, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         return this.breakNaturally(null);
     }
 
@@ -511,6 +556,11 @@ public class CraftBlock implements Block {
 
     @Override
     public boolean applyBoneMeal(BlockFace face) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, this.position, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         Direction direction = blockFaceToNotch(face);
         Preconditions.checkArgument(direction != null, face + " is not a valid cartesian face");
 
@@ -531,8 +581,10 @@ public class CraftBlock implements Block {
             capturedBlockStates = new ArrayList<>(world.capturedBlockStates.values());
             world.capturedBlockStates.clear();
 
-            treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
+            // Pufferfish start - Parallel world ticking
+            treeType = SaplingBlock.getTreeTypeTL();
+            SaplingBlock.setTreeTypeTL(null);
+            // Pufferfish end - Parallel world ticking
         }
 
         if (!capturedBlockStates.isEmpty()) {
@@ -620,6 +672,11 @@ public class CraftBlock implements Block {
 
     @Override
     public RayTraceResult rayTrace(Location start, Vector direction, double maxDistance, FluidCollisionMode fluidCollisionMode) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         Preconditions.checkArgument(start != null, "Location start cannot be null");
         Preconditions.checkArgument(this.getWorld().equals(start.getWorld()), "Location start cannot be a different world");
         start.checkFinite();
@@ -661,6 +718,11 @@ public class CraftBlock implements Block {
 
     @Override
     public boolean canPlace(BlockData data) {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThreadOrAsyncThread(serverWorld, "Cannot read world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         Preconditions.checkArgument(data != null, "BlockData cannot be null");
 
         BlockState state = ((CraftBlockData) data).getState();
@@ -699,6 +761,11 @@ public class CraftBlock implements Block {
 
     @Override
     public void tick() {
+        // Pufferfish start - Parallel world ticking
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && this.level instanceof ServerLevel serverWorld) {
+            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(serverWorld, this.position, "Cannot modify world asynchronously");
+        }
+        // Pufferfish end - Parallel world ticking
         final ServerLevel level = this.level.getMinecraftWorld();
         this.getBlockState().tick(level, this.position, level.getRandom());
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
index 1ef1943c9934400e40c3c060299c98659709a5cc..a5ae2ca3b96d116669e218cf4a16f483a80ed90a 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockEntityState.java
@@ -34,6 +34,25 @@ public abstract class CraftBlockEntityState<T extends BlockEntity> extends Craft
     private final T snapshot;
     public boolean snapshotDisabled; // Paper
     public static boolean DISABLE_SNAPSHOT = false; // Paper
+    // Pufferfish start - Parallel world ticking
+    public static ThreadLocal<Boolean> DISABLE_SNAPSHOT_TL = ThreadLocal.withInitial(() -> Boolean.FALSE);
+
+    public static boolean getDisableSnapshotTL() {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking && DISABLE_SNAPSHOT_TL.get()) return true;
+
+        synchronized (CraftBlockEntityState.class) {
+            return DISABLE_SNAPSHOT;
+        }
+    }
+
+    public static void setDisableSnapshotTL(boolean value) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) DISABLE_SNAPSHOT_TL.set(value);
+
+        synchronized (CraftBlockEntityState.class) {
+            DISABLE_SNAPSHOT = value;
+        }
+    }
+    // Pufferfish end - Parallel world ticking
 
     public CraftBlockEntityState(World world, T blockEntity) {
         super(world, blockEntity.getBlockPos(), blockEntity.getBlockState());
@@ -42,8 +61,10 @@ public abstract class CraftBlockEntityState<T extends BlockEntity> extends Craft
 
         try { // Paper - Show blockstate location if we failed to read it
         // Paper start
-        this.snapshotDisabled = DISABLE_SNAPSHOT;
-        if (DISABLE_SNAPSHOT) {
+        // Pufferfish start - Parallel world ticking
+        this.snapshotDisabled = getDisableSnapshotTL();
+        if (snapshotDisabled) {
+        // Pufferfish end - Parallel world ticking
             this.snapshot = this.blockEntity;
         } else {
             this.snapshot = this.createSnapshot(blockEntity);
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
index ec3343ecbe68b1c2a33f85edb3b527363affa9ba..3e988712d88fcf0adb1d7031f5027af3646cf3fc 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
@@ -346,6 +346,7 @@ public class CraftBlockState implements BlockState {
 
     @Override
     public java.util.Collection<org.bukkit.inventory.ItemStack> getDrops(org.bukkit.inventory.ItemStack tool, org.bukkit.entity.Entity entity) {
+        if (gg.pufferfish.pufferfish.PufferfishConfig.enableParallelWorldTicking) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(world.getHandle(), position, "Cannot modify world asynchronously"); // Pufferfish - Parallel world ticking
         this.requirePlaced();
         net.minecraft.world.item.ItemStack item = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(tool);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
index 063a4ad33d1c12ec2cc1dff5239a7bf58b92fabb..87506169a91e97a9c35941f9f179cc16a03a3fac 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
@@ -198,14 +198,16 @@ public final class CraftBlockStates {
         BlockPos pos = craftBlock.getPosition();
         net.minecraft.world.level.block.state.BlockState state = craftBlock.getBlockState();
         BlockEntity blockEntity = craftBlock.getLevel().getBlockEntity(pos);
-        boolean prev = CraftBlockEntityState.DISABLE_SNAPSHOT;
-        CraftBlockEntityState.DISABLE_SNAPSHOT = !useSnapshot;
+        // Pufferfish start - Parallel world ticking
+        boolean prev = CraftBlockEntityState.getDisableSnapshotTL();
+        CraftBlockEntityState.setDisableSnapshotTL(!useSnapshot);
+        // Pufferfish end - Parallel world ticking
         try {
             CraftBlockState blockState = CraftBlockStates.getBlockState(world, pos, state, blockEntity);
             blockState.setWorldHandle(craftBlock.getLevel()); // Inject the block's level accessor
             return blockState;
         } finally {
-            CraftBlockEntityState.DISABLE_SNAPSHOT = prev;
+            CraftBlockEntityState.setDisableSnapshotTL(prev); // Pufferfish - Parallel world ticking
         }
     }
 
